File: 1D DP/000 Pracitce/Coin Change/main.py
class Solution:
    @staticmethod
    def coin_change(coins: list[int], target: int):
        coins.sort()
        dp = [float("inf")] * (target+1)
        dp[0] = 0

        for i in range(1, target+1):
            for coin in coins:
                if coin > i: break

                dp[i] = min(
                    dp[i-coin] + 1,
                    dp[i]
                )

        return dp[target] if dp[target] != float("inf") else -1


print(Solution.coin_change(
    coins = [25, 10, 5, 30], target = 30
))

File: 1D DP/000 Pracitce/Coin Change/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution{
    public:
    int coin_change(vector<int>& coins, int amount){
        sort(coins.begin(), coins.end());

        int n = coins.size();
        vector<int> dp(amount+1, amount+1);
        dp[0] = 0;

        for (int i=1; i<amount+1; i++){
            for(int coin: coins){
                
                if (coin > i) break;

                dp[i] = min(
                    dp[i-coin]+1,
                    dp[i]
                );
            }
        }
        if (dp[amount] == amount+1) return -1;
        else return dp[amount];
    }
};


int main(){
    Solution solution;
    vector<int> coins = {474,83,404,3};
    int amount = 264;
    cout << solution.coin_change(coins, amount);
    return 0;
}